cpp/src/neighbors/detail/nn_descent.cuh

```diff
@@ -1057,14 +1064,17 @@ void GnndGraph<Index_t>::init_random_graph()
       size_t base_idx      = i * node_degree + seg_idx * segment_size;
       auto h_neighbor_list = h_graph + base_idx;
       auto h_dist_list     = h_dists.data_handle() + base_idx;
+      size_t idx           = base_idx;
       for (size_t j = 0; j < static_cast<size_t>(segment_size); j++) {
-        size_t idx = base_idx + j;
         Index_t id = rand_seq[idx % rand_seq.size()] * num_segments + seg_idx;
         if ((size_t)id == i) {
-          id = rand_seq[(idx + segment_size) % rand_seq.size()] * num_segments + seg_idx;
+          idx = small_dataset ? (idx + 1) : (idx + segment_size);
```

Comment:
Hi @cjnolet, @achirkin , @anaruse @tfeher , the second issue I'm trying to resolve here, particularly in a small dataset where the dataset size is less than 32 (the segment size), is minimizing the likelihood of self-nodes in the initial graph.

Reply:
In the previous indexing, `id` was dependent on the loop variable `j`, this is not the case anymore. Therefore I do not think this is correct.

I do not understand why this would help.

Reply:
It partially worked because the duplicated nodes will not be unique in the sequent code, which will cause the duplicated node in the output graph in small dataset scenarios. Now, I know this change is not the last resolution, and it only resolves a super small dataset problem in the new cases added. But the conclusion is not changed: I'm pretty sure this `init_random_graph` has a bug, because it assumes the `nrow >> node_degree`:
```
std::vector<Index_t> rand_seq(nrow / num_segments); // when nrow=34, num_segments=2, the 2 segments will repeat in a `0-17`, full of repeated nodes as neighbors.
```
. I'm trying to provide the right change.

Reply:
The result of an improper initial graph will affect the consequence sample process and the last output graph.